hoomd/md/AlchemicalPotentialPair.h

```diff
@@ -0,0 +1,385 @@
+        {
+        // TODO: make sure only adding a particle to an alchemostat can enable it
+        // TODO: is this where the momentum etc should be initilized?
+        m_alchemy_mask[m_alchemy_index(alpha_p->m_type_pair_param.x, alpha_p->m_type_pair_param.y)]
+                      [alpha_p->m_type_pair_param.z]
+            = true;
+        alpha_p->resizeForces(this->m_pdata->getN());
+        }
+
+    void disableAlchemicalPairParticle(std::shared_ptr<AlchemicalPairParticle> alpha_p)
+        {
+        // TODO: handle resetting param value and alpha
+        m_alchemy_mask[m_alchemy_index(alpha_p->m_type_pair_param.x, alpha_p->m_type_pair_param.y)]
+                      [alpha_p->m_type_pair_param.z]
+            = false;
+        }
+
+    // TODO: keep as general python object, or change to an updater? problem with updater is lack of
+    // input, but could make a normalizer based on it
+    void setNormalizer(std::function<Scalar(pybind11::kwargs)>& callback)
+        {
+        m_normalized = true;
+        m_normalizer = callback;
+        }
+
+    // TODO: disable alchemical particle by resetting params, mask, should remain shared_ptr?
+
+    protected:
+    typedef std::bitset<evaluator::num_alchemical_parameters> mask_type;
+    typedef std::array<Scalar, evaluator::num_alchemical_parameters> alpha_array_t;
+
+    Index2DUpperTriangular m_alchemy_index; //!< upper triangular typepair index
+    std::vector<mask_type> m_alchemy_mask;  //!< Type pair mask for if alchemical forces are used
+    std::vector<std::shared_ptr<AlchemicalPairParticle>>
+        m_alchemical_particles; //!< 2D array (alchemy_index,alchemical param)
+    // static constexpr bool m_normalized =
+    // std::is_member_function_pointer<decltype(evaluator::setNormalizationValue)>::value;    //!<
+    // The potential should be normalizeds
+
+    // pybind11::function m_normalizer;
+    // template<typename ...Args> std::function<Scalar(Args...)> m_normalizer;
+    std::function<Scalar(pybind11::kwargs)> m_normalizer;
+    bool m_normalized = false;
+
+    //! Method to be called when number of types changes
+    void slotNumTypesChange() override;
+    void slotNumParticlesChange()
+        {
+        unsigned int N = this->m_pdata->getN();
+        for (auto& particle : m_alchemical_particles)
+            if (particle)
+                particle->resizeForces(N);
+        };
+
+    // Extra steps to insert
+    inline AlchemyPackage<evaluator> pkgInitialize(const uint64_t& timestep) override;
+    inline void pkgPerNeighbor(const unsigned int& i,
+                               const unsigned int& j,
+                               const unsigned int& typei,
+                               const unsigned int& typej,
+                               const bool& in_rcut,
+                               evaluator& eval,
+                               AlchemyPackage<evaluator>&) override;
+    inline void pkgFinalize(AlchemyPackage<evaluator>&) override;
```

Comment:
The code to be injected at various points related to the alchemical information. Hopefully general enough it could be applied to other use cases if needed in the future. Unless it prevents a GPU implementation.

Reply:
I will rewrite as a subclass. I would prefer minimal code duplication in a subclass over unused template parameters for the majority of PotentialPair.

At this time, there are no plans for a GPU implementation.